### PR TITLE
[Serialization] Give swiftdocs a stable version

### PR DESCRIPTION
**Pull request by jrose-apple**
We're committing to this as a forwards-compatible format, and in most cases probably backwards-compatible as well! Break it apart from the AST/SIL swiftmodule format, give it its own version number, and…probably not have any problems, because there's nothing complicated here.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
Could you provide more context why do we need this version number while it's both forward and backward compatible?

**Comment by jrose-apple** [comment]
`SWIFTDOC_VERSION_MINOR` isn't really important for forwards/backwards-compatibility, but it gives you the option to conditionalize behavior based on which version of the compiler generated the file. (For example, maybe in a newer format, an absent record indicates that a decl should be omitted from the generated interface, while in the older format it just means it has no comment.)

`SWIFTDOC_VERSION_MAJOR` is important in case you really do have to break compatibility some day. I suspect it won't actually go up at all, but if it does then how the rest of the file is processed can be conditionalized on the major version number.

This format is basically [semantic versioning](https://semver.org), but a file format isn't exactly the same thing as an API, so I didn't mention semver by name in the doc comment.

**Comment by nkcsgexi** [comment]
Thanks for the clarification! Can we derive `SWIFTDOC_VERSION_MINOR` from the complier/module format version we're using rather than maintaining it separately? Does this change also imply that all tooling invocation should also specify `SWIFTDOC_VERSION_MAJOR`?

File: lib/Serialization/DocFormat.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/// describe what change you made. The content of this comment isn't important;
+/// it just ensures a conflict if two people change the module format.
+/// Don't worry about adhering to the 80-column limit for this line.
+const uint16_t SWIFTDOC_VERSION_MINOR = 1; // Last change: skipping 0 for testing purposes
+
+/// The record types within the comment block.
+///
+/// Be very careful when changing this block; it must remain stable. Adding new
+/// records is okay---they will be ignored---but modifying existing ones must be
+/// done carefully. You may need to update the version when you do so.
```
Comment (akyrtzi) [inline lib/Serialization/DocFormat.h]: What is the effect of adding a new 'field' at the end of an existing record, could you clarify it in the comments ?
Can we determine how many fields a record has at deserialization so we don't have to update and check for a version number ?

File: lib/Serialization/DocFormat.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+/// Increment this value when making a backwards-incompatible change, which
+/// should be rare. When incrementing this value, reset SWIFTDOC_VERSION_MINOR
+/// to 0.
+const uint16_t SWIFTDOC_VERSION_MAJOR = 1;
+
+/// Serialized swiftdoc format minor version number.
+///
+/// Increment this value when making a backwards-compatible change that might
+/// be interesting to test for. However, if old swiftdoc files are fully
+/// compatible with the new change, you do not need to increment this.
```
Comment (akyrtzi) [inline lib/Serialization/DocFormat.h]: I'm probably reading this paragraph wrong but it seems to contradict itself, could you rephrase for clarify, are you making a distinction between "backwards-compatible" and "fully compatible" ? If yes I'd recommend to define in details (and maybe with some examples) what you mean by each.

File: lib/Serialization/DocFormat.h (hunk 1 above)
Comment (jrose-apple) [reply lib/Serialization/DocFormat.h]: You can kind of do that—see https://github.com/apple/swift/blob/bae3964c07a976ee0978c0c2400e23e3252e60e8/lib/Serialization/ModuleFile.cpp#L200-L216—but BCRecordLayout doesn't support it. You have to read records manually in that case. Adding new records isn't really that expensive, though, so I don't think it's too much of a problem to do that.

For a stable format, the version number is supposed to be higher-level than just "did I change something". If it makes no difference, it's fine not to update it (and if you really don't want to use it, we could just drop the "minor" version entirely).

**Comment by jrose-apple** [comment]
> Thanks for the clarification! Can we derive `SWIFTDOC_VERSION_MINOR` from the complier/module format version we're using rather than maintaining it separately?

Unfortunately no. The goal here is to have a version number for a format that's stable across compiler versions. The binary swiftmodule format definitely isn't.

> Does this change also imply that all tooling invocation should also specify `SWIFTDOC_VERSION_MAJOR`?

I'm not quite sure what you're asking here. The way the version checks are written, all our tools are compiled with exactly one expected major version for any swiftdoc they read. If that version ever changes, we'll have to decide how the new tools handle the differing major versions, but the old tools will correctly reject the newer generated files.

(It's also theoretically possible to decide what version to use when *generating* the file—say, if a user doesn't use any complicated features—but that's probably not something you'll need, especially since the way your deserialization is written will already ignore records it doesn't understand.)

**Comment by akyrtzi** [comment]
@jrose-apple thanks so much for doing this! My main concern is for `DocFormat.h` to contain documentation with crystal clear instructions and directions on how to make changes and maintain compatibility, it should have no ambiguities.

File: lib/Serialization/DocFormat.h (hunk 1 above)
Comment (akyrtzi) [reply lib/Serialization/DocFormat.h]: I think it's good that you added the minor version, as an extra thing we may find useful to check,  I'm mainly trying to understand if we should update it when adding a new field.

Adding a new field would force us to switch to fully manually reading the record ? Would we be able to do a check and switch to using a different BCRecordLayout to deserialize that record, and could this check occur without needing to update the minor version ?

Pointing in the documentation to the example that you gave on how to manually check for the number of fields would be useful.

**Comment by nkcsgexi** [comment]
> I'm not quite sure what you're asking here. The way the version checks are written, all our tools are compiled with exactly one expected major version for any swiftdoc they read. If that version ever changes, we'll have to decide how the new tools handle the differing major versions, but the old tools will correctly reject the newer generated files.

ah, so the versioning is entirely transparent to compiler clients like sourcekitd, and its purpose is to separate the evolution of doc-serialization with compiler and module format because in practice it's way more stable, right?

File: lib/Serialization/DocFormat.h (hunk 1 above)
Comment (jrose-apple) [reply lib/Serialization/DocFormat.h]: It's unfortunately worse than that: adding a new field will **break existing tools**, because BCRecordLayout currently asserts that the number of fields matches exactly. https://github.com/apple/swift-llvm/blob/stable/include/llvm/Bitcode/RecordLayout.h#L261

We could change that, of course, or add a new API `readIgnoringTrailingFields` or something, but the point is that you can only add new fields if there aren't existing tools expecting a fixed number of fields.

**Comment by jrose-apple** [comment]
> ah, so the versioning is entirely transparent to compiler clients like sourcekitd, and its purpose is to separate the evolution of doc-serialization with compiler and module format because in practice it's way more stable, right?

Yep, that sounds right to me!

> My main concern is for `DocFormat.h` to contain documentation with crystal clear instructions and directions on how to make changes and maintain compatibility, it should have no ambiguities.

Agreed. I'll add more comments based on the discussions we're having now.

**Comment by jrose-apple** [comment]
@akyrtzi @nkcsgexi How does this look?

@swift-ci Please smoke test

**Review by nkcsgexi — APPROVED** [review]

**Review by akyrtzi — APPROVED** [review]
Love the detailed documentation!

**Comment by jrose-apple** [comment]
@swift-ci Please test
